### PR TITLE
Add tests for namespaces and acl bootstrapping

### PR DESCRIPTION
**Pull request by lkysow**
Add tests for the namespace changes of server-acl-init.

The tests that will fail are due to the binding rule not being able to have its namespace updated.

Tests:
* binding rule
  * single dest namespace
  * mirroring
* auth method
  * single dest namespace
  * mirroring
* updates (i.e. when the command is re-run it updates the binding rule/auth method)
  * no ns => ns mirroring
  * no ns => single dest namespace <= FAILS
  * ns mirroring => single dest namespace <= FAILS
  * single dest namespace => ns mirroing

### CODE REVIEW

File: subcommand/server-acl-init/command_ent_test.go [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+
+	bootToken := getBootToken(t, k8s, resourcePrefix)
+	consul, err := api.NewClient(&api.Config{
+		Address: testAgent.HTTPAddr,
+		Token:   bootToken,
+	})
+	require.NoError(err)
+
+	// Check the ACL auth method is created in the expected namespace.
+	authMethodName := "release-name-consul-k8s-auth-method"
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: It would be good to use the `releaseName` variable to construct this so it works if `releaseName` changes.

File: subcommand/server-acl-init/command_ent_test.go [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+	methods, _, err := consul.ACL().AuthMethodList(&api.QueryOptions{
+		Namespace: "dest",
+	})
+	require.NoError(err)
+	found := false
+	for _, m := range methods {
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: Alternatively, there should only be one auth method in the `dest` namespace, so you could check that the length of the list is 1 and that the single item is the correct auth method.

File: subcommand/server-acl-init/command_ent_test.go [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+		if m.Name == authMethodName {
+			found = true
+		}
+	}
+	require.True(found, fmt.Sprintf("no auth method with name %s found", authMethodName))
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: There should be some validation about the properties of the auth method here, since they matter and can vary with namespace enabled.

Specifically, it would be good to check its Config: `MapNamespaces` and `ConsulNamespacePrefix` fields.

File: subcommand/server-acl-init/command_ent_test.go [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+	cmd.init()
+	args := []string{
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-k8s-namespace=" + ns,
+		"-expected-replicas=1",
+		"-create-inject-auth-method",
+		"-enable-namespaces",
+		"-enable-inject-k8s-namespace-mirroring",
+		"-inject-k8s-namespace-mirroring-prefix=prefix-",
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: There should be another test case with an empty prefix. Also, it would be good to test that providing a `-consul-inject-destination-namespace` field gets overridden by the mirroring.

File: subcommand/server-acl-init/command_ent_test.go [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+	bootToken := getBootToken(t, k8s, resourcePrefix)
+	consul, err := api.NewClient(&api.Config{
+		Address: testAgent.HTTPAddr,
+		Token:   bootToken,
+	})
+	require.NoError(err)
+
+	// Check the ACL auth method is as expected.
+	authMethodName := "release-name-consul-k8s-auth-method"
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: Same comment about using the `releaseName` variable to construct this.

File: subcommand/server-acl-init/command_ent_test.go [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+	cases := map[string]struct {
+		// Args for first run of command.
+		FirstRunArgs []string
+		// Args for second run of command.
+		SecondRunArgs []string
+		// Expected namespace for the auth method.
+		AuthMethodExpectedNS string
+		// Whether to expect the MapNamespaces and ConsulNamespacePrefix
+		// keys to be set on the config.
+		AuthMethodExpectMirroringConfig bool
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: The mirroring and prefix checking should be separate and the prefix fields validated.

File: subcommand/server-acl-init/command_ent_test.go [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+		// Expected namespace for the binding rule.
+		BindingRuleExpectedNS string
+	}{
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: Additional test cases should check that changing the mirroring prefix gets updated. That includes empty string to non-empty, non-empty to empty and non-empty to non-empty with different values.

File: subcommand/server-acl-init/command_ent_test.go [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+				"-server-label-selector=component=server,app=consul,release=" + releaseName,
+				"-resource-prefix=" + resourcePrefix,
+				"-k8s-namespace=" + ns,
+				"-expected-replicas=1",
+				"-create-inject-auth-method",
+			}
+
+			// First run. NOTE: we don't assert anything here since we've
+			// tested these results in other tests. What we care about here
+			// is the result after the second fun.
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: I like fun, but I don't think that's what you meant :).
```suggestion
			// is the result after the second run.
```

File: subcommand/server-acl-init/command_ent_test.go [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+			cmd := Command{
+				UI:        ui,
+				clientset: k8s,
+			}
+			responseCode := cmd.Run(append(defaultArgs, c.FirstRunArgs...))
+			require.Equal(0, responseCode, ui.ErrorWriter.String())
+
+			// Second run.
+			cmd = Command{
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: Does this need to be declared again?

File: subcommand/server-acl-init/command_ent_test.go [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+			// Now check that everything is as expected.
+			bootToken := getBootToken(t, k8s, resourcePrefix)
+			consul, err := api.NewClient(&api.Config{
+				Address: testAgent.HTTPAddr,
+				Token:   bootToken,
+			})
+			require.NoError(err)
+
+			// Check the ACL auth method is as expected.
+			authMethodName := "release-name-consul-k8s-auth-method"
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: `releaseName` again.

File: subcommand/server-acl-init/command_ent_test.go (hunk 7 above)
Comment (adilyse) [reply subcommand/server-acl-init/command_ent_test.go]: It would also be good to have one that tests a single destination namespace that is the Consul `default` namespace.

File: subcommand/server-acl-init/command_ent_test.go [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+	t.Parallel()
+	k8s, testAgent := completeEnterpriseSetup(t, resourcePrefix)
+	defer testAgent.Stop()
+	require := require.New(t)
+
+	ui := cli.NewMockUi()
+	firstRunArgs := []string{
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-k8s-namespace=" + ns,
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: Besides adding an update scenario to going to mirroring, it would also be good to test the case where the k8s namespace is not `default`.

File: subcommand/server-acl-init/command_ent_test.go [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+		"-create-client-token",
+		"-allow-dns",
+		"-create-mesh-gateway-token",
+		"-create-sync-token",
+		"-create-inject-namespace-token",
+		"-expected-replicas=1",
+		"-allow-dns",
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: This is declared twice.
```suggestion
```

File: subcommand/server-acl-init/command_ent_test.go [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+				require.NoError(err)
+				require.NotContains(policy.Rules, "namespace")
+				break
+			}
+		}
+		require.True(found, "Did not find "+expected)
+	}
+
+	// Re-run the command with namespace flags. The policies should be updated.
+	cmd = Command{
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: Does this need to be defined again?

File: subcommand/server-acl-init/command_ent_test.go [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+
+	ui := cli.NewMockUi()
+	firstRunArgs := []string{
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-k8s-namespace=" + ns,
+		"-create-client-token",
+		"-allow-dns",
+		"-create-mesh-gateway-token",
+		"-create-sync-token",
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: It would be good to include the snapshot agent and ent license policy creation as well so that we can validate that they don't change with namespaces.

File: subcommand/server-acl-init/command_ent_test.go [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+		"client-token",
+		"catalog-sync-token",
+		"connect-inject-token",
+		"mesh-gateway-token",
+	}
+	policies, _, err := consul.ACL().PolicyList(nil)
+	require.NoError(err)
+	for _, expected := range expectedPolicies {
+		found := false
+		for _, p := range policies {
```
Comment (adilyse) [inline subcommand/server-acl-init/command_ent_test.go]: What about doing a single loop through `policies` first to put them in a map of policy name -> policy struct? Then when looping through `expectedPolicies`, you can look them up directly.

File: subcommand/server-acl-init/command_ent_test.go [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+	"github.com/mitchellh/cli"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	"testing"
+)
+
+// Test that ACL policies get updated if namespaces config changes.
+func TestRun_ACLPolicyUpdates(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (adilyse) [reply subcommand/server-acl-init/command_ent_test.go]: Also, it might be good to test the `createOrUpdateACLPolicy` method directly.

File: subcommand/server-acl-init/command_ent_test.go [hunk 17]
```diff
@@ -0,0 +1,436 @@
+// +build enterprise
+
+package serveraclinit
+
+import (
+	"encoding/base64"
+	"fmt"
+	"github.com/hashicorp/consul/api"
```
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: I think we should get some ci goodness for this

File: subcommand/server-acl-init/command_ent_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: > Also, it might be good to test the createOrUpdateACLPolicy method directly.

It looks like the only thing we're not testing is this:
```
		// This shouldn't happen, because we're looking for a policy
		// only after we've hit a `Policy already exists` error.
		if policy.ID == "" {
			return errors.New("Unable to find existing ACL policy")
		}
```

If that's true then not sure this warrants a separate test unless there's a case I'm missing.

File: subcommand/server-acl-init/command_ent_test.go (hunk 11 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Do you mean testing the policy updates from when the kubernetes namespace is not default or another type of update? AFAIK the k8s namespace doesn't affect the policies at all right?

File: subcommand/server-acl-init/command_ent_test.go (hunk 12 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: 🍺

File: subcommand/server-acl-init/command_ent_test.go (hunk 15 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Good call, switched to:

```go
	// Collect the actual polices into a map to make it easier to assert
	// on their existence and contents.
	actualPolicies := make(map[string]string)
	for _, p := range policies {
		policy, _, err := consul.ACL().PolicyRead(p.ID, nil)
		require.NoError(err)
		actualPolicies[p.Name] = policy.Rules
	}
	for _, expected := range expectedPolicies {
		actPolicy, ok := actualPolicies[expected]
		require.True(ok, "Did not find policy %s", expected)
		// We assert that the policy doesn't have any namespace config
		// in it because later that's what we're using to test that it
		// got updated.
		require.NotContains(actPolicy, "namespace")
	}
```

File: subcommand/server-acl-init/command_ent_test.go (hunk 13 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Yes 😄found out about this the hard way. If we re-use the old command then old flags are still set. In this case it won't make a difference but if the first-run had set `-destination-consul-ns` then even though you set `-mirror-k8s` in the second run it would never take effect since the old flag would still be set.

In this case, it won't have effect but I think it's better to have it re-defined for two reasons:
1. If the test changes in the future it's very confusing to debug
1. Its a better representation of the real-world use-case. If a user upgrades with helm the command won't be re-used, it will be redefined since it's a brand new job. So we wouldn't want a side-effect in this case to cause the test to pass.

File: subcommand/server-acl-init/command_ent_test.go (hunk 1 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: 🍺

File: subcommand/server-acl-init/command_ent_test.go (hunk 9 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: See comment above. In this case it has to be otherwise the old flags stay set.

File: subcommand/server-acl-init/command_ent_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: > It makes more sense to me to have the initial functionality tests before the update tests.

Down to move it but I'll do it after you have a second look so the changes are easier to review.

File: subcommand/server-acl-init/command_ent_test.go (hunk 14 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Done.

File: subcommand/server-acl-init/command_ent_test.go (hunk 2 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Switched to
```go
			// Check the ACL auth method is created in the expected namespace.
			authMethodName := releaseName + "-consul-k8s-auth-method"
			namespaceQuery := &api.QueryOptions{
				Namespace: consulDestNamespace,
			}
			actMethod, _, err := consul.ACL().AuthMethodRead(authMethodName, namespaceQuery)
			require.NotNil(actMethod)
			require.Equal("kubernetes", actMethod.Type)
			require.Equal("Kubernetes AuthMethod", actMethod.Description)
			require.NotContains(actMethod.Config, "MapNamespaces")
			require.NotContains(actMethod.Config, "ConsulNamespacePrefix")

			// Check the binding rule is as expected.
			rules, _, err := consul.ACL().BindingRuleList(authMethodName, namespaceQuery)
			require.NoError(err)
			require.Len(rules, 1)
			actRule, _, err := consul.ACL().BindingRuleRead(rules[0].ID, namespaceQuery)
			require.NoError(err)
			require.NotNil(actRule)
			require.Equal("Kubernetes binding rule", actRule.Description)
			require.Equal(api.BindingRuleBindTypeService, actRule.BindType)
			require.Equal("${serviceaccount.name}", actRule.BindName)
			require.Equal("serviceaccount.name!=default", actRule.Selector)

```

File: subcommand/server-acl-init/command_ent_test.go (hunk 3 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Yep, done!

File: subcommand/server-acl-init/command_ent_test.go (hunk 4 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Added cases:
```go
	cases := map[string]struct {
		MirroringPrefix string
		ExtraFlags      []string
	}{
		"no prefix": {
			MirroringPrefix: "",
			ExtraFlags:      nil,
		},
		"with prefix": {
			MirroringPrefix: "prefix-",
			ExtraFlags:      nil,
		},
		"with destination namespace flag": {
			MirroringPrefix: "",
			// Mirroring takes precedence over this flag so it should have no
			// effect.
			ExtraFlags: []string{"-consul-inject-destination-namespace=dest"},
		},
	}

```

File: subcommand/server-acl-init/command_ent_test.go (hunk 6 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Changed to
```
		// If true, we expect MapNamespaces to be set on the auth method
		// config.
		AuthMethodExpectMapNamespacesConfig bool
		// If AuthMethodExpectMapNamespacesConfig is true, we will assert
		// that the ConsulNamespacePrefix field on the auth method config
		// is set to this.
		AuthMethodExpectedNamespacePrefixConfig string

```

File: subcommand/server-acl-init/command_ent_test.go (hunk 7 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: > Additional test cases should check that changing the mirroring prefix gets updated. That includes empty string to non-empty, non-empty to empty and non-empty to non-empty with different values.

Added
* mirroring ns (no prefix) => mirroring ns (no prefix)
* mirroring ns => mirroring ns (same prefix)
* mirroring ns (no prefix) => mirroring ns (prefix)
* mirroring ns (prefix) => mirroring ns (no prefix)

> It would also be good to have one that tests a single destination namespace that is the Consul default namespace.

I added cases to `TestRun_ConnectInject_SingleDestinationNamespace`: `consulDestNamespaces := []string{"default", "destination"}`. How's that?

File: subcommand/server-acl-init/command_ent_test.go [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+			rules, _, err := consul.ACL().BindingRuleList(authMethodName, &api.QueryOptions{
+				Namespace: c.BindingRuleExpectedNS,
+			})
+			require.NoError(err)
+			require.Len(rules, 1)
+		})
+	}
+}
+
+func setupK8sServiceAccount(t *testing.T, k8s *fake.Clientset) {
```
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Done! Good catch. This method had to move to `command_test` though because in this file it won't get picked up by the OSS tests due to the build tag.

File: subcommand/server-acl-init/command_ent_test.go (hunk 7 above)
Comment (adilyse) [reply subcommand/server-acl-init/command_ent_test.go]: Looks good!

File: subcommand/server-acl-init/command_ent_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Yeah for sure, I did change it but it reverted. I've set my editor now so should be good in the future.

File: subcommand/server-acl-init/command_ent_test.go (hunk 11 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: So we both agree that this flag has no effect on whether the policies are updated and we've already tested the templating of the policies in the rules_test.go. It sounds like you're saying the reason we should test this flag varying is because it's related to namespaces. I don't think that's a good enough reason to add an additional case here. You could make the same argument for the -allow-k8s-namespace flag. In the end I think that would complicate the test without a corresponding gain in meaningful testing coverage.

File: subcommand/server-acl-init/command_ent_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: I think those cases are already covered through the tests. I ran with code coverage:
image

File: subcommand/server-acl-init/command_ent_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Okay I double checked this time, I think we're good.

File: subcommand/server-acl-init/command_ent_test.go (hunk 13 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: done.

File: subcommand/server-acl-init/command_ent_test.go [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+				Token:   bootToken,
+			})
+			require.NoError(err)
+
+			// Check the ACL auth method is created in the expected namespace.
+			authMethodName := releaseName + "-consul-k8s-auth-method"
+			namespaceQuery := &api.QueryOptions{
+				Namespace: consulDestNamespace,
+			}
+			actMethod, _, err := consul.ACL().AuthMethodRead(authMethodName, namespaceQuery)
```
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: done (lines 174-178)

**Comment by lkysow** [comment]
> Did you get a chance to add a test for being able to update the value of the binding rule? It's not enterprise specific, but that functionality changed as part of the namespace work so it would be good to add one. 

Sorry I totally forgot that from your original review. I've added `TestRun_BindingRuleUpdates` to `command_test.go`.

I also forgot your comment from before:
> Targeted tests for the new checkAndCreateNamespace method

I looked at code coverage and it's fully tested except for the simple error cases:

image

I like that we're mostly testing the public interfaces here while getting full coverage. In the case of a particularly complicated helper method I'm down to write separate tests but looking at the existing code coverage this seems okay to me.

File: subcommand/server-acl-init/command_ent_test.go (hunk 11 above)
Comment (adilyse) [reply subcommand/server-acl-init/command_ent_test.go]: It's a test against future human error, where someone modifies the wrong namespace flag.

File: subcommand/server-acl-init/command_ent_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (adilyse) [reply subcommand/server-acl-init/command_ent_test.go]: This is an argument for doing comprehensive testing at the lowest possible level rather than up the call tree somewhere. Test organization rather than code coverage.

File: subcommand/server-acl-init/command_test.go [hunk 20]
```diff
@@ -406,6 +407,63 @@ func TestRun_ConnectInjectAuthMethod(t *testing.T) {
...
+		"-acl-binding-rule-selector=serviceaccount.name!=changed",
+	)
+
+	// Run the command first to populate the binding rule.
+	cmd := Command{
+		UI:        ui,
+		clientset: k8s,
+	}
+	responseCode := cmd.Run(firstRunArgs)
+	require.Equal(0, responseCode, ui.ErrorWriter.String())
```
Comment (adilyse) [inline subcommand/server-acl-init/command_test.go]: This needs a check that the binding rule matched the first argument after the first run.

Comment (lkysow) [reply subcommand/server-acl-init/command_test.go]: In these update tests the pattern has been to only check that it changed. The first run creation is tested in other tests. It was done this way to reduce the amount of test code. If it breaks for the first run, it will be detected in the other tests. I've added it though.

File: subcommand/server-acl-init/command_test.go [hunk 21]
```diff
@@ -406,6 +407,63 @@ func TestRun_ConnectInjectAuthMethod(t *testing.T) {
...
+	firstRunArgs := []string{
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-k8s-namespace=" + ns,
+		"-expected-replicas=1",
+		"-create-inject-auth-method",
+		"-acl-binding-rule-selector=serviceaccount.name!=default",
+	}
+	// Our second run, we change the binding rule selector.
+	secondRunArgs := append(firstRunArgs,
```
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_test.go]: Yeah good call, moved to

```go
	commonArgs := []string{
		"-server-label-selector=component=server,app=consul,release=" + releaseName,
		"-resource-prefix=" + resourcePrefix,
		"-k8s-namespace=" + ns,
		"-expected-replicas=1",
		"-create-inject-auth-method",
	}
	firstRunArgs := append(commonArgs,
		"-acl-binding-rule-selector=serviceaccount.name!=default",
	)
	// Our second run, we change the binding rule selector.
	secondRunArgs := append(commonArgs,
		"-acl-binding-rule-selector=serviceaccount.name!=changed",
	)
```

File: subcommand/server-acl-init/command_ent_test.go (hunk 11 above)
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Okay I added it.

File: subcommand/server-acl-init/command_ent_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (lkysow) [reply subcommand/server-acl-init/command_ent_test.go]: Yes, I prefer testing the public interfaces and allowing the private implementation to vary. In my experience it makes the code easier to work with because you can refactor without having to update a ton of tests. It's a balance tradeoff in case your public interface tests get super complicated to exercise corner cases in helper methods. In my judgement it's not worth it in this case.
